codalab/lib/completers.py

```diff
@@ -1,39 +1,84 @@
-'''
+"""
 completers.py
-'''
+"""
 import inspect
 import itertools
+import os
+import re
 
 from argcomplete import warn
 from argcomplete.completers import FilesCompleter
 
+
 from codalab.lib import spec_util, worksheet_util
 
+
+KEY_TARGET_SPEC_FORMAT = r"(?:([^:]*):)?([^/]*)(?:/(.*))?"
+GLOBAL_WORKSHEET_SPEC_FORMAT = r"(?:(.*)::)?(.*)"
+
+
 def short_uuid(full_uuid):
     return worksheet_util.apply_func('[0:8]', full_uuid)
 
+
+def require_not_headless(completer):
+    """
+    Given a completer, return a CodaLabCompleter that will only call the
+    given completer if the client is not headless.
+    """
+    class SafeCompleter(CodaLabCompleter):
+        def __call__(self, *args, **kwargs):
+            if self.cli.headless:
+                return ()
+            elif inspect.isclass(completer):
```

Comment:
Do we really need to multiplex between a class and an instance?  Why not have the caller add '()' if necessary?